File: book/section8/generics.hs
data Foo a = Person {age :: Int, name :: String, something :: a}

x :: Foo Int
x = Person {age = 31, name = "nishio", something = 1}

y :: Foo String
y = Person {age = 31, name = "nishio", something = "hoge"}

main = do
  print $ something x   -- -> 1
  print $ something y   -- -> "hoge"

  

File: langbook/chapter08/generics.cpp
#include <iostream>

template<typename T>
struct person {
  int age;
  char *name;
  T something;
};

int main(){
  person<int> x;
  x.something = 1;
  person<const char*> y;
  y.something = "hoge";

  std::cout << x.something << std::endl; // -> 1
  std::cout << y.something << std::endl; // -> hoge
}
